worlds/lingo/static_logic.py

```diff
@@ -197,363 +70,49 @@ def get_progressive_item_id(name: str):
-    else:
-        req_blocked = False
-
-    if "req_blocked_when_no_doors" in painting_data:
-        req_blocked_when_no_doors = painting_data["req_blocked_when_no_doors"]
-    else:
-        req_blocked_when_no_doors = False
-
-    required_door = None
-    if "required_door" in painting_data:
-        door = painting_data["required_door"]
-        required_door = RoomAndDoor(
-            door["room"] if "room" in door else room_name,
-            door["door"]
-        )
-
-    painting_obj = Painting(painting_id, room_name, enter_only, exit_only, orientation,
-                            required_painting, rwnd, required_door, disable_painting, move_painting, req_blocked,
-                            req_blocked_when_no_doors)
-    PAINTINGS[painting_id] = painting_obj
-    PAINTINGS_BY_ROOM[room_name].append(painting_obj)
-
-
-def process_progression(room_name, progression_name, progression_doors):
-    global PROGRESSIVE_ITEMS, PROGRESSION_BY_ROOM
-
-    # Progressive items are configured as a list of doors.
-    PROGRESSIVE_ITEMS.append(progression_name)
-
-    progression_index = 1
-    for door in progression_doors:
-        if isinstance(door, Dict):
-            door_room = door["room"]
-            door_door = door["door"]
-        else:
-            door_room = room_name
-            door_door = door
-
-        room_progressions = PROGRESSION_BY_ROOM.setdefault(door_room, {})
-        room_progressions[door_door] = Progression(progression_name, progression_index)
-        progression_index += 1
-
-
-def process_room(room_name, room_data):
-    global ROOMS, ALL_ROOMS
-
-    room_obj = Room(room_name, [])
-
-    if "entrances" in room_data:
-        for source_room, doors in room_data["entrances"].items():
-            process_entrance(source_room, doors, room_obj)
-
-    if "panels" in room_data:
-        PANELS_BY_ROOM[room_name] = dict()
-
-        for panel_name, panel_data in room_data["panels"].items():
-            process_panel(room_name, panel_name, panel_data)
-
-    if "doors" in room_data:
-        DOORS_BY_ROOM[room_name] = dict()
-
-        for door_name, door_data in room_data["doors"].items():
-            process_door(room_name, door_name, door_data)
-
-    if "paintings" in room_data:
-        PAINTINGS_BY_ROOM[room_name] = []
-
-        for painting_data in room_data["paintings"]:
-            process_painting(room_name, painting_data)
-
-    if "progression" in room_data:
-        for progression_name, progression_doors in room_data["progression"].items():
-            process_progression(room_name, progression_name, progression_doors)
+def load_static_data_from_file():
+    global HASHES, PAINTINGS, ALL_ROOMS, DOORS_BY_ROOM, PANELS_BY_ROOM, PROGRESSIVE_ITEMS, PROGRESSION_BY_ROOM, PAINTING_ENTRANCES, PAINTING_EXIT_ROOMS, PAINTING_EXITS, REQUIRED_PAINTING_ROOMS, REQUIRED_PAINTING_WHEN_NO_DOORS_ROOMS, SPECIAL_ITEM_IDS, PANEL_LOCATION_IDS, DOOR_LOCATION_IDS, DOOR_ITEM_IDS, DOOR_GROUP_ITEM_IDS, PROGRESSIVE_ITEM_IDS
```

Comment:
:D

I would probably change all these to `HASHES.update(pickdata["HASHES"])` (or .extend in the case of a list, I believe), in which case the global stuff isn't necessary

The other thing you could do is just have this function return pickdata itself, and then do the assignments at the top level. You run into some weird stuff though I guess if you get the "return None" case. Could do something like

```
pickdata = load_static_data_from_file()  # This returns pickdata itself now
if pickdata is None:
    pickdata = defaultdict(key: lambda=dict())

ALL_ROOMS = list(pickdata["ALL_ROOMS"])
DOORS_BY_ROOM = pickdata["DOORS_BY_ROOM"]
PANELS_BY_ROOM = pickdata["PANELS_BY_ROOM"]
```

Idk probably just do the "update" thing. Or leave it if you really want, global is just one of those keywords you're trained as a Python dev to have a visceral repulsive reaction to, and in this case the line length violation doesn't make it better :D

Reply:
Two of these were primitives, so I had to keep them global, but I changed the rest of them.

Reply:
That's fair